angularSTEP/src/app/upload-recipe/upload-recipe.component.ts

```diff
@@ -149,21 +222,72 @@ export class UploadRecipeComponent{
     this.formatInstructions(this.instructionsArray);
     this.extraFormGroup.controls.extraInfo.setValue(this.autoCapitalizeFirst(this.extraFormGroup.value.extraInfo));
 
-    this.db.collection('recipes')
-      .ref.withConverter(new RecipeConverter().recipeConverter)
-      .add(new Recipe(
-        this.basicsFormGroup.value.name,
-        '',
-        this.basicsFormGroup.value.difficulty,
-        this.basicsFormGroup.value.description,
-        this.ingredientsArray.value,
-        this.toolsArray.value,
-        this.instructionsArray.value,
-        this.extraFormGroup.value.extraInfo ? this.extraFormGroup.value.extraInfo : '',
-        Date.now(),
-        [], 
-        [],
-    ));
-    this.router.navigate(['/confirm-upload']);
+    if (this.branching) {
+      const ingredientsCheck = (currentValue: string) => this.ingredientsArray.value.indexOf(currentValue) === this.baseRecipe.ingredients.indexOf(currentValue);
+      const toolsCheck = (currentValue: string) => this.toolsArray.value.indexOf(currentValue) === this.baseRecipe.tools.indexOf(currentValue);
+      const instructionsCheck = (currentValue: string) => this.instructionsArray.value.indexOf(currentValue) === this.baseRecipe.instructions.indexOf(currentValue);
+  
+      if (
+        this.basicsFormGroup.controls.name.value === this.baseRecipe.recipeName
+        &&  this.basicsFormGroup.controls.difficulty.value === this.baseRecipe.difficulty
+        &&  this.basicsFormGroup.controls.description.value === this.baseRecipe.description
+        &&  this.ingredientsArray.value.every(ingredientsCheck)
+        &&  this.toolsArray.value.every(toolsCheck)
+        &&  this.instructionsArray.value.every(instructionsCheck)
+        &&  this.extraFormGroup.controls.extraInfo.value === this.baseRecipe.extraInfo
+     ) {
+        this.changed = false;
+      } else {
+        this.changed = true;
+
+        this.fAuth.currentUser.then(user => {
+          if (user) {
+            this.db.collection('recipes')
+            .ref.withConverter(new RecipeConverter().recipeConverter)
+            .add(new Recipe(
+              this.basicsFormGroup.value.name, 
+              user.uid,
+              this.basicsFormGroup.value.difficulty,
+              this.basicsFormGroup.value.description,
+              this.ingredientsArray.value,
+              this.toolsArray.value,
+              this.instructionsArray.value,
+              this.extraFormGroup.value.extraInfo ? this.extraFormGroup.value.extraInfo : '',
+              Date.now(),
+              [], 
+              [],
+              this.id,
+              this.baseUploaderUid,
+            )).then(() => {
+              this.router.navigate(['/confirm-upload']);
+            });
+          }
+        });
+      }
+    } else {
```

Comment:
You've got basically the same code repeated. Could you do the following? In general, it's good to keep code as unindented as possible, since layers of conditionals tend to cause more bugs.

if (branching)
  if (<nothing has changed>) {
    this.changed = false;
    return;
  } else {
    this.changed = true;
 }
<add recipe>